### ISSUE TITLE
Super Weird issues re transactions and connection pools

### ISSUE THREAD

**Issue by grigi** [tortoise/tortoise-orm#40]
I have tried to enable connection pools as part of the test suite, but there is a very very odd issue with connection pools and transactions.

When using a single connection, you always get the same connection, so we don't have to worry about which connection is busy in what transaction, but with a pool, one needs to lock the connection for the duration of the connection.

It looks like it should, but it isnt.

Symptoms I have found so far are:
1. using a different connection than the one that begun the transaction
2. some operations, such as `.update` use the connection very differently than `.filter` Why do we have code sections that reserve a single connection? and some that dont?
3. A subsequent query in a transaction doesn't see the prior changes made in the transaction
4. postgresql just hangs around and waits for something

Things I have tried:
1. investigated the aiocontext, it seems to work?
2. Tried making the get_single_connection of BaseExecutor.execute_select behave like the update, which just calls a query on the given client (like update does)
3. Tried making update do a get_single_connection like most things in BaseExecutor

This all leaves me quite confused as to why we have so many different ways of wrapping connections, and why they just don't seem to work with pools.

I want a clear idea before I actually consider ripping things apart.

**Comment by Zeliboba5** [comment]
Hi
Currently I can't go all out investigating it, will look further into in on weekend.
From what I managed to see right now:

> postgresql just hangs around and waits for something

Problem seems to be that connections in pool ends and test hangs up waiting for connection to be released. So the problem is why they are not released by previous tests?

> some operations, such as .update use the connection very differently than .filter Why do we have code sections that reserve a single connection? and some that dont?

The thing is `.filter` doesn't make a query itself, it just prepares values for it. Actual query is made by executor class, and it was meant that all execute methods in executor should reserve single connection for all of a query. You can see that it is reserved in `execute_select` the same way it is reserved in `execute_update`. That was made exactly because of problems with pool I had in the initial stage of project, when recursive operations just took all of the connections from pool and it hang up the same way, waiting for connection to be released, never aquiring it, because upper level query was waiting for recursive queries to be done.

I think bug should be somewhere there, may be some part of code doesn't honour this idea and just takes fresh connection from the pool, so because of it `subsequent query in a transaction doesn't see the prior changes made in the transaction`.

Also - may be problem with `green` running multiply tests concurrently and different tests hang up on middle because all connections are already taken, but even if it is this way, the bug is still there because query should be able to finish itself within one connection and releasing it after finish

**Comment by grigi** [comment]
I already checked `green` and single process, definitely not that.

Yes, I was talking of how `execute_select` reserved a single connection, and how `UpdateQuery` does not.
Also the whole transaction set was re-worked recently, and it is likely that due to changes in behaviour there we broke pooling (or rather I broke it unwittingly).

I feel that the grabbing of a single connection seems... over-complicating things. Likely we don't understand the real issue?
We probably need tests that operate on the backends directly, so we can determine if the problem is in our client implementation, or the executor, or the queryset?
